tests/unit/test_charm.py

```diff
@@ -1,23 +1,374 @@
+
+        requirers_certificates = self.harness.charm.certificates_requirers.get_issued_certificates(
+            requirer_relation_id
+        )
+        self.assertEqual(len(requirers_certificates["certificates-requirer"]), 0)
+
+    def test_given_provided_certificate_when_certificate_expired_then_event_is_ignored(
+        self,
+    ) -> None:
+        self._integrate_provider()
+        requirer_relation_id = self._integrate_requirer()
+        self.harness.update_relation_data(
+            requirer_relation_id,
+            "certificates-requirer/0",
+            key_values={"certificate_signing_requests": get_json_csr_list()},
+        )
+        available_event = CertificateAvailableEvent(
+            handle=Mock(),
+            certificate_signing_request="test_csr",
+            certificate="test_cert",
+            ca="test_ca",
+            chain=["test_ca", "test_intermediate"],
+        )
+        self.harness.charm._on_certificate_available(event=available_event)
+
+        requirers_certificates = self.harness.charm.certificates_requirers.get_issued_certificates(
+            requirer_relation_id
+        )
+        self.assertEqual(len(requirers_certificates["certificates-requirer"]), 1)
+
+        invalidated_event = CertificateInvalidatedEvent(
+            handle=Mock(),
+            reason="expired",
+            certificate_signing_request="test_csr",
+            certificate="test_cert",
+            ca="test_ca",
+            chain=["test_ca", "test_intermediate"],
+        )
+        self.harness.charm._on_certificate_invalidated(event=invalidated_event)
+
+        requirers_certificates = self.harness.charm.certificates_requirers.get_issued_certificates(
+            requirer_relation_id
+        )
+        self.assertEqual(len(requirers_certificates["certificates-requirer"]), 1)
+
+    def test_given_multiple_provided_certificate_when_all_certificates_invalidated_then_all_certificates_are_removed(  # noqa: E501
+        self,
+    ) -> None:
+        provider_relation_id = self._integrate_provider()
+        requirer1_relation_id = self._integrate_requirer()
+        self.harness.update_relation_data(
+            requirer1_relation_id,
+            "certificates-requirer/0",
+            key_values={"certificate_signing_requests": get_json_csr_list("test_csr1")},
+        )
+        available_event = CertificateAvailableEvent(
+            handle=Mock(),
+            certificate_signing_request="test_csr1",
+            certificate="test_cert1",
+            ca="test_ca",
+            chain=["test_ca", "test_intermediate"],
+        )
+        self.harness.charm._on_certificate_available(event=available_event)
+        requirer2_relation_id = self._integrate_requirer("certificates-requirer2")
+        self.harness.update_relation_data(
+            requirer2_relation_id,
+            "certificates-requirer2/0",
+            key_values={"certificate_signing_requests": get_json_csr_list("test_csr2")},
+        )
```

Comment:
I can remove most of the unit tests in that case, because they will not test anything of value if we mock the lib.

Using mocks here will reduce the value of the tests. The only gain I can see is a bit of test code setting relation data.

Reply:
I am comfortable with removing most of the unit test code here. We should be able to do so while keeping the same coverage. The unit under test is our charm code. In the context of certificate invalidated event, we have a 1 liner that calls the lib, it makes sense to have a extremely simple test that checks the call with the correct data.

Reply:
I disagree. The call to the library is an implementation detail of our charm. I doubt this will change, but it still is an implementation detail. I think it makes sense to mock out calls to the filesystem, databases and other external dependencies. However, I do not think we should mock things that stay inside our process. We do not call mocks to a bunch of other libraries that we use all the time, like `json`, `logging` and `jinja2`. We do mock calls to the filesystem, the k8s apis, etc.

Mocks come with a big cost. They make test more brittle because they inevitably couple the test to the implementation. We should use them as little as possible.

Reply:
Why are we having tests that call `self.harness.charm._on_certificate_invalidated(event=invalidated_event)` in this case instead of simply manipulating relation data?

Following your suggestion will highly complexify all our unit tests. They will all have to manipulate charm relation data and effectively replicate the behavior that is already being owned and tested by our libs.

Reply:
This is the reason I was suggesting having something in the lib to help with test. Let's say a method like `event_to_relation_data`. It would let us easily trigger the real events and have more meaningful tests, without having to know all the implementation details.

With `json` for example, we write tests like this for something that receives JSON and constructs objects from it:

```python
def test():
    data = json.dumps([{"test": "data"}])

    result = unit_under_test(data)

    assert result == Test("data")
```

We do not write:

```python
@patch("json.loads")
def test(json_mock):
    json_mock.return_value = Test("data")

    result = unit_under_test(Mock())

    assert result == Test("data")
```

(I know, this example is extreme, but you get the point)

In this example, the behavior for everything related to JSON is owned and tested by the `json` builtin library, yet it still gets exercised in our tests. The test in my first example needs to change for only 2 reasons: the input of the unit under test changes, or the expected output changes. The second one also needs to change for those same reasons, but needs to also change when we change the implementation, let's say to using another json library, or the json library itself changes a bit in how you call it, etc.

You are right about the tests currently in this PR that directly call the event handler. I think the tests would be better if I exercised the public interface of our charm (the relation data). I think we should make that easy to do instead of using mocks.

Reply:
Agreed, I sent a meeting invite to the whole team to have this discussion.